### ISSUE TITLE
onChange providing the clicked option

### ISSUE THREAD

**Issue by oyeanuj** [benbowes/react-responsive-select#142]
Hi @benbowes, thanks for creating this great library!

As I was integrating it, the only thing I'm having a hard time with is how `onChange` works. It seems that `onChange` passes all the selected options rather than the one selected/deselected. This requires a tedious diff every time to find which option changed. It seems like the library already knows that, so is there a way to receive that information?

(And please let me know if I'm missing something which is already there that I can use)!

**Comment by bbowes-ae** [comment]
Hi @oyeanuj thank for the PR earlier.

Are you talking about `multiselect` mode?

These are the onchange broadcaster params:
https://github.com/benbowes/react-responsive-select/blob/master/src/lib/onChangeBroadcasters/multiSelectBroadcastChange.ts#L5-L8

Are you talking about passing back the current and previous options? Or an intersection of what has changed?

I'm after just a bit more detail.

Thanks for the input 👍

**Comment by oyeanuj** [comment]
@benbowes Appreciate the quick response! 

I was referring to the intersection of what has changed or even making it behave like a standard input by the passing the event that causes the change, so  one knows what option was selected/deselected. ([Here is what `react-super-select` does for example](http://alsoscotland.github.io/react-super-select/#onChange)).

I imagine calling an user-provided event handler here (and `onHandleEnterPressed`) could help: https://github.com/benbowes/react-responsive-select/blob/70d169a6c15631e5f8f3cb90ca0d28e7d592dbbf/src/lib/eventHandlers/handleClick.ts#L40-L49

**Comment by oyeanuj** [comment]
(Just updated the comment above with another idea and more clarified thoughts)

**Comment by benbowes** [comment]
I'm toying with the idea of doing a diff and passing the result back as a second argument of the `onChange` prop.

Currently trying this:

```
import { IOutputMultiSelect, IOutputMultiSelectOption } from "../../types/";
import { isEqual } from "../../lib/isEqual";

type Diff = {
  added?: IOutputMultiSelectOption[];
  removed?: IOutputMultiSelectOption[];
}

function difference(a: any = [], b: any = []): Diff {
  const added = a.filter(
    (avalue: any) =>
      !b.some(
        (bvalue: any) => JSON.stringify(bvalue) === JSON.stringify(avalue)
      )
  );
  const removed = b.filter((avalue: any) =>
    a.every((bvalue: any) => JSON.stringify(bvalue) !== JSON.stringify(avalue))
  );

  return added.length > 0 ? { added } : { removed };
}

export function multiSelectBroadcastChange(
  currOptions: IOutputMultiSelectOption[],
  altered: boolean,
  fn?: (
    changes: IOutputMultiSelect,
    diff?: Diff,
  ) => void,
  prevOptions?: IOutputMultiSelectOption[]
): void {
  if (!fn) return;

  const shouldBroadcastChange =
    !prevOptions || !isEqual(prevOptions, currOptions);

  if (shouldBroadcastChange) {
    const diff: Diff = difference(currOptions, prevOptions);

    fn(
      {
        options: currOptions.map((currOption: IOutputMultiSelectOption) => ({
          name: currOption.name || "",
          text: currOption.text || "",
          value: currOption.value || "",
        })),
        altered,
      },
      diff
    );
  }
}
```

That would return an object as the second argument of either:
- { added: [{ name: '', text: '', value: '' }] }
- { removed: [{ name: '', text: '', value: '' }] }

There is no native event to pass back.

**Comment by benbowes** [comment]
`onSelect` and `onDeselect` might be easier implement and use though 🤔

**Comment by oyeanuj** [comment]
@benbowes I agree that `onSelect` and `onDeselect` might be easier. They also might be more helpful as it allows people to do other things on those events as well, like closing the select or some other side effect.

So, I assume if one uses `onSelect`/`onDeselect`, effectively, we won't need to use `onChange`, right?

**Comment by benbowes** [comment]
Yes, leave `onChange` as is, and add `onSelect/onDeselect` sounds good to me

**Comment by benbowes** [comment]
So do you want me to make this change? Or are you planning to do it?

**Comment by benbowes** [comment]
I had play and it seems to work well. Pre

**Comment by benbowes** [comment]
Closing accidental

**Comment by benbowes** [comment]
@oyeanuj I have made a pr for this https://github.com/benbowes/react-responsive-select/pull/143

**Comment by benbowes** [comment]
see https://github.com/benbowes/react-responsive-select/pull/150

**Comment by benbowes** [comment]
released in 6.0.0
